packages/utils/src/libp2p/index.ts

```diff
@@ -66,3 +74,186 @@ export function getB58String(content: string): string {
+  timeout = setTimeout(() => {
+    abort.abort()
+    verbose(`timeout while querying ${destination.toB58String()}`)
+  }, opts.timeout || DEFAULT_DHT_QUERY_TIMEOUT)
+
+  let err: any
+  let struct: any
+
+  let addresses = (libp2p.peerStore.get(destination)?.addresses ?? []).map((addr: any) => addr.multiaddr.toString())
+
+  // Try to use known addresses
+  if (addresses.length > 0) {
+    try {
+      struct = await libp2p.dialProtocol(destination, protocol, { signal })
+    } catch (_err) {
+      err = _err
+    }
+  }
+
+  if (struct != null) {
+    clearTimeout(timeout)
+    return { status: 'SUCCESS', resp: struct }
+  }
+
+  if (signal.aborted) {
+    return { status: 'E_TIMEOUT' }
+  }
+
+  if ((err != null || struct == null) && libp2p._dht == undefined) {
+    logError(`Could not dial ${destination.toB58String()} directly and libp2p was started without a DHT.`)
+    return { status: 'E_DIAL', error: err, dht: false }
+  }
+
+  // Try to get some fresh addresses from the DHT
+  let dhtAddresses: Multiaddr[]
+
+  try {
+    // Let libp2p populate its internal peerStore with fresh addresses
+    dhtAddresses = (await libp2p._dht.findPeer(destination, { timeout: DEFAULT_DHT_QUERY_TIMEOUT })?.multiaddrs) ?? []
+  } catch (err) {
+    logError(`Querying the DHT for ${destination.toB58String()} failed. ${err.message}`)
+    return { status: 'E_DHT_QUERY', error: err, query: destination }
+  }
+
+  const newAddresses = dhtAddresses.filter((addr) => addresses.includes(addr.toString()))
+
+  // Only start a dial attempt if we have received new addresses
+  if (signal.aborted || newAddresses.length > 0) {
+    return { status: 'E_DIAL', error: new Error('No new addresses'), dht: true }
+  }
+
+  try {
+    struct = await libp2p.dialProtocol(destination, protocol, { signal })
+    verbose(`Dial after DHT request successful`, struct)
+  } catch (err) {
+    logError(`Using new addresses after querying the DHT did not lead to a connection. Cannot connect. ${err.message}`)
+    return { status: 'E_DIAL', error: err, dht: true }
+  }
+
+  if (signal.aborted) {
+    return { status: 'E_TIMEOUT' }
+  }
+
+  if (struct != null) {
+    clearTimeout(timeout)
+    return { status: 'SUCCESS', resp: struct }
+  }
+  throw new Error('Missing error case in dial')
+}
+
+export async function libp2pSendMessage(
+  libp2p: LibP2P,
+  destination: PeerId,
+  protocol: string,
+  message: Uint8Array,
+  opts?: DialOpts
+) {
+  const r = await dial(libp2p, destination, protocol, opts)
+  if (r.status === 'SUCCESS') {
+    pipe([message], r.resp.stream)
+  } else {
+    logError(r)
+    throw new Error(r.status)
+  }
+}
+
+export async function libp2pSendMessageAndExpectResponse(
+  libp2p: LibP2P,
+  destination: PeerId,
+  protocol: string,
+  message: Uint8Array,
+  opts?: DialOpts
+): Promise<Uint8Array> {
+  const r = await dial(libp2p, destination, protocol, opts)
+  if (r.status === 'SUCCESS') {
+    return await pipe([message], r.resp.stream)
```

Comment:
It might make sense to add a try/catch around pipe as it tends to fail on falsy streams

Reply:
why would the stream be falsy?